### PR TITLE
Add `smart_open` dependency to `ray[default]`

### PR DESCRIPTION
**Pull request by shrekris-anyscale**
## Why are these changes needed?

In feature #20280, support for remote dependencies accessed via remote URIs was expanded to HTTPS and GS. However, all remote URIs require the user to `pip install smart_open` to use them. Rather than forcing the user to manually download this dependency, this code change allows users to download `smart_open` by running `pip install ray[default]`.

These changes add `smart_open` as a default Ray dependency, and they include minimal dependency tests to ensure that Ray can continue to make progress without `smart_open` when using locally-hosted dependencies. Additionally, they ensure that when Ray does throw an error for situations which require `smart_open`, the error contains a message instructing the user to `pip install smart_open`.

## Related issue number

Closes issue #20358.

## Checks

- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - Two minimal dependency tests have been added to `test_runtime_env_ray_minimal.py`. These tests confirm that Ray can use locally-hosted dependencies without `smart_open` and that Ray provides useful messages when it does error on logic that requires `smart_open`.

### CODE REVIEW

File: python/ray/tests/test_runtime_env_ray_minimal.py [hunk 1]
```diff
@@ -103,5 +106,70 @@ def test_ray_client_init(call_ray_start):
...
+    if option == "working_dir":
+        env = {"working_dir": uri}
+    elif option == "py_modules":
+        env = {"py_modules": [uri]}
+
+    try:
+        with pytest.raises(Exception) as err:
+            ray.init(address, runtime_env=env)
+        assert "pip install smart_open" in str(err.value)
+    except Failed:
```
Comment (architkulkarni) [inline python/ray/tests/test_runtime_env_ray_minimal.py:134]: Why do we need this try-except?  The underscore in front of `_pytest` makes me think it's some internal module we shouldn't need to use, but I haven't seen it before so I'm not sure

**Comment by architkulkarni** [comment]
@richardliaw could you please give codeowner approval / dependencies approval?

File: python/ray/tests/test_runtime_env_ray_minimal.py (hunk 1 above)
Comment (edoakes) [reply python/ray/tests/test_runtime_env_ray_minimal.py:134]: +1, shouldn't be using internal types like this

Comment (edoakes) [reply python/ray/tests/test_runtime_env_ray_minimal.py:134]: I'm not actually sure what this except clause is doing -- do we always expect it to not raise an exception?

**Review by edoakes — APPROVED** [review]
@richardliaw LGTM'd offline as well

**Comment by architkulkarni** [comment]
Now that smart_open has been added as a ray[default] dependency, it looks like the test is failing with a different error message before it gets to the smart_open message: "RuntimeError: Not all required Ray dependencies for the runtime_env feature were found on the cluster. To install the required dependencies, please run `pip install "ray[default]"` on all cluster nodes."

This error message is correct, and it provides the correct solution, so maybe it's okay to get rid of the newly added tests in this PR?  To use runtime envs at all, `ray[default]` needs to be installed.

**Comment by architkulkarni** [comment]
Because runtime envs requires ray[default] and we're adding smart_open to ray[default], the following behavior 

"Ensures that runtime_env logic
    still supports uploading local packages to GCS, even when smart_open is
    not installed."

will no longer be supported.

**Comment by shrekris-anyscale** [comment]
Yeah that's true. We can rework the minimal dependency tests, so they check for the new error message too.

The main change is that we no longer support GCS uploads without `smart_open`. I agree with @architkulkarni that since we need `ray[default]` anyways for runtime environments, and since `smart_open` is a part of runtime environments, this should OK.

@edoakes Are you OK with the change as well?

**Comment by edoakes** [comment]
@shrekris-anyscale @architkulkarni yes, the changes SGTM

**Review by jiaodong — APPROVED** [review]

**Pull request merged**